lib/logflare/gen_singleton.ex

```diff
@@ -0,0 +1,91 @@
+defmodule Logflare.GenSingleton do
+  @moduledoc """
+  A generic singleton GenServer that will be unique cluster-wide.
+  Checks if there the server is started, if not, will be started under the supervision tree as a transient GenServer.
+  """
+  use GenServer
+
+  @default_check 5_000
+
+  require Logger
+  @type state :: any()
+
+  @spec start_link(args :: any()) :: {:ok, pid} | {:error, any}
+  def start_link(args) do
+    GenServer.start_link(__MODULE__, args, [])
+  end
+
+  def count_children(pid) do
+    GenServer.call(pid, :count_children)
+  end
+
+  @impl true
+  def init(args) do
+    interval = args[:interval] || @default_check
+    {:ok, pid} = Supervisor.start_link([], strategy: :one_for_one)
```

Comment:
You typically never want to start a supervisor inside another process (except another supervisor). This is because otherwise you have to reimplement the shutdown semantics. So idea here is to have GenSingleton be a supervisor that starts the GenServer. And then the GenServer can start the `child_spec` as a sibling in the supervision tree.

The tricky here is for the GenServer to capture the supervisor by passing self on `start_link`:

```elixir
  def start_link(args) do
    # Self here is the supervisor PID
    GenServer.start_link(__MODULE__, {self(), args}, [])
  end
```

Reply:
ah very nice didn't think of that 👍